plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/SystemMismatchDetector.java

```diff
@@ -0,0 +1,274 @@
+        Objects.requireNonNull(entry, "Expected entry.");
+        logger.debug("Entry '" + entry.plantDescription() + "' added, checking for inconsistencies...");
+        updateAlarms();
+    }
+
+    @Override
+    public void onPlantDescriptionUpdated(
+        final PlantDescriptionEntry updatedEntry,
+        final PlantDescriptionEntry oldEntry
+    ) {
+        Objects.requireNonNull(updatedEntry, "Expected entry.");
+        logger.debug("Entry '" + updatedEntry.plantDescription() + "' updated, checking for inconsistencies...");
+        updateAlarms();
+    }
+
+    @Override
+    public void onPlantDescriptionRemoved(final PlantDescriptionEntry entry) {
+        Objects.requireNonNull(entry, "Expected entry.");
+        logger.debug("Entry '" + entry.plantDescription() + "' removed, checking for inconsistencies...");
+        updateAlarms();
+    }
+
+    @Override
+    public void onSystemAdded(final SrSystem system) {
+        Objects.requireNonNull(system, "Expected system.");
+        logger.debug("System '" + system.systemName() + "' added, checking for inconsistencies...");
+        updateAlarms();
+    }
+
+    @Override
+    public void onSystemRemoved(final SrSystem system) {
+        Objects.requireNonNull(system, "Expected system.");
+        logger.debug("System '" + system.systemName() + "' removed, checking for inconsistencies...");
+        updateAlarms();
+    }
+
+    /**
+     * @param entrySystem      A system in a Plant Description Entry.
+     * @param registeredSystem A system retrieved from the Service registry.
+     * @return True if the two objects represent the same real-world system,
+     * false otherwise.
+     */
+    private boolean systemsMatch(final PdeSystem entrySystem, final SrSystem registeredSystem) {
+
+        if (!Metadata.isSubset(entrySystem.metadata(), registeredSystem.metadata())) {
+            return false;
+        }
+
+        // At this point, we know that either of these two statements must hold:
+        // a) The entry system has metadata which matches that of the SR system.
+        // b) The entry system has no metadata, in which case it *must* have a
+        //    systemName (as per the spec).
+
+        if (entrySystem.systemName().isPresent()) {
+            final String entryName = entrySystem.systemName().get();
+            return entryName.equals(registeredSystem.systemName());
+        }
+
+        return true;
+    }
+
+    /**
+     * @param alarm  An alarm.
+     * @param system A system retrieved from the Service registry.
+     * @return True if the alarm refers to the given system, false otherwise.
+     */
+    private boolean alarmMatchesSrSystem(final Alarm alarm, final SrSystem system) {
+
+        final String alarmSystemName = alarm.getSystemName();
+        final Map<String, String> alarmMetadata = alarm.getMetadata();
+        final boolean namesMatch = alarmSystemName != null && alarmSystemName.equals(system.systemName());
```

Comment:
My previous comments may be relevant here too.

Reply:
System names are now validated at the time they enter the PDE.